### ISSUE TITLE
Add streamlined API

### ISSUE THREAD

**Issue by darribas** [pysal/mapclassify#72]
Opening up a thread there to sketch together what a different, streamlined API could look like for `mapclassify` algorithms. This grows out of discussions with other projects (e.g. [`xarray`](https://github.com/makepath/xarray-spatial/issues/30), [`ipyleaflet`](https://github.com/jupyter-widgets/ipyleaflet/issues/261). The basic idea is we would like `mapclassify` to be as easy to use and useful as possible to potential users, and we've started thinking maybe our current API could be streamlined for the "80% of cases" where  you have a 1D array with values and you want back a set of labels.

Based on this, we were thinking of starting by adding a method that'd wrap around all the available methods available for classification. For a very rough sketch, something like:

```python
a = numpy.random.random(100)
q = mapclassify.classify(a, "quantiles", k=5)
```

Another option would be to develop it following the `sklearn` pattern:

```python
classifier = mapclassify.Quantiles(k=5)
q = classifier.fit_transform(a)
```

And there might be others more useful for folks. Please if you feel inclined, do drop your views here, it'd be really useful to get as many views as possible. Also, of course, any other ideas or "wish-list" items you may have that'd make you more likely to use/adopt `mapclassify` for choropleth classification tasks, we'd love to help if possible.

Tagging a few folks we imagine might be interested in some way (@brendancol, @kristinepetrosyan, @martinRenou, @jorisvandenbossche, @sjsrey, @ljwolf, @slumnitz ), but feel free to tag more as you go along!

**Comment by sjsrey** [comment]
Adding @ResidentMario

**Comment by slumnitz** [comment]
option 1 is already implemented in [splot here](https://github.com/pysal/splot/blob/2e8f13a54e71faa60c0520ea9ba7e2b30aad2629/splot/_viz_value_by_alpha_mpl.py#L400) and could be moved.

**Comment by sjsrey** [comment]
> Opening up a thread there to sketch together what a different, streamlined API could look like for `mapclassify` algorithms. This grows out of discussions with other projects (e.g. [`xarray`](https://github.com/makepath/xarray-spatial/issues/30), [`ipyleaflet`](https://github.com/jupyter-widgets/ipyleaflet/issues/261). The basic idea is we would like `mapclassify` to be as easy to use and useful as possible to potential users, and we've started thinking maybe our current API could be streamlined for the "80% of cases" where you have a 1D array with values and you want back a set of labels.
> 
> Based on this, we were thinking of starting by adding a method that'd wrap around all the available methods available for classification. For a very rough sketch, something like:
> 
> ```python
> a = numpy.random.random(100)
> q = mapclassify.classify(a, "quantiles", k=5)

is `q` _just_ the labels, or a classifier object?

**Comment by darribas** [comment]
My sense was that `q` would be a 1D array with the labels from the classification. But again, this is more to open up the discussion than anything else.

**Comment by sjsrey** [comment]
@darribas something like:

![image](https://user-images.githubusercontent.com/118042/83975924-77bd7380-a8ab-11ea-8385-5e2dac712066.png)


This implementation would return the instance, thinking that the user may also want the intervals (and possibly other attributes). In the case they wanted the labels, they could chain the attribute access (cell [5]).

**Comment by martinfleis** [comment]
👍 

This could simplify the implementation in GeoPandas as some of the checks we do there, would likely be done in `mapclassify.classify`. I like @sjsrey's proposal and its ability to return `bins` and potentially `k` alongside labels.

**Comment by brendancol** [comment]
@darribas This all looks great...just chiming in with my humble opinion: 

I would prefer the following spelling...`mapclassify.<classification_method>(data, hyperparams) -> bin_edges`

```python
a = numpy.random.random(100)
bins = mapclassify.quantiles(a, k=5)
```

where `type(bins)` is np.ndarray of bin edges. I care less about the bin_ids because I might be classifying on a subset of data.

**Comment by sjsrey** [comment]
> @darribas This all looks great...just chiming in with my humble opinion:
> 
> I would prefer the following spelling...`mapclassify.<classification_method>(data, hyperparams) -> bin_edges`
> 
> ```python
> a = numpy.random.random(100)
> bins = mapclassify.quantiles(a, k=5)
> ```
> 
> where `type(bins)` is np.ndarray of bin edges. I care less about the bin_ids because I might be classifying on a subset of data.

Something like this is already in there:
```python
>>> import numpy
>>> import mapclassify
>>> x = numpy.random.random(100)
>>> bins = mapclassify.classifiers.quantile(x, k=5)
>>> bins
array([0.20836597, 0.36675624, 0.61913714, 0.7582157 , 0.99642854])
```
We could refactor the namespace and provide the binning functions for all the classifiers. In other words, to change:
```python
>>> bins = mapclassify.classifiers.quantile(x, k=5)
```
to
```python
>>> bins = mapclassify.quantile(x, k=5)
```
and do this for all classifiers.

**Comment by slumnitz** [comment]
@sjsrey, @darribas, @martinfleis what would be the con's if we supported multiple API's?

**Comment by martinfleis** [comment]
@slumnitz multiple API's are always coming with additional maintenance cost and can be a bit of a barrier for new users (more things to understand).

**Comment by slumnitz** [comment]
Thank you @martinfleis for your thoughts! Just to put it out there for discussion and brainstorming: Since @sjsrey and @darribas `mapclassify.classify` suggestion is already supported by PySAL there might not be a big add on in maintenance cost if we move the functionality from splot here.

In terms of barriers to users, we could emphasise one of the methods for researchers/ users i.e. `mapclassify.classify(y)` and simply keep the other one, if needed for developers `mapclassify.classifiers.quantile()`, but not promote is heavily in documentation to make clear one of the solutions should be the 'main' one. (Similarly to i.e. `pandas.read_hdf` and then for `pandas.HDFStore` which has more functionality.) 

I am not sure though if this would apply well to `mapclassify` as both APIs would basically provide the same functionality? Just curious to know what others think? Should we settle on one? Or might there be 'lower level' or 'higher level functionality' that could be worth including?

**Comment by martinfleis** [comment]
In this particular case, I feel that the ideal long-term situation is to have `mapclassify.classify(y, scheme='Quantiles')` as a primary API and keeping all `mapclassify.classifiers.quantile()`-like as backend. Because that is what essentially GeoPandas does and most of the other software. You have a single API where you choose which scheme you want. 

I fully agree with @slumnitz here in terms of emphasizing streamlined API in docs. I see it as a lower-level and higher-level API. Yes, both come with the same functionality, but since one will be likely used in the other, the maintenance cost will not be as high as the user experience gained. But I'd prefer if both had the same functions, i.e. same methods and attributes.

**Comment by ljwolf** [comment]
I think, too, that there should be a `return_labels` argument that turns the return into a tuple of (bins, labels), like numpy.unique

**Comment by slumnitz** [comment]
Very cool! :) Something to discuss on the next dev meeting?

**Comment by martinfleis** [comment]
Closed by #90
